core/src/main/java/com/graphhopper/routing/util/spatialrules/AreaIndex.java

```diff
@@ -0,0 +1,79 @@
+/*
+ *  Licensed to GraphHopper GmbH under one or more contributor
+ *  license agreements. See the NOTICE file distributed with this work for
+ *  additional information regarding copyright ownership.
+ *
+ *  GraphHopper GmbH licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except in
+ *  compliance with the License. You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+// todo: eventually move to another package
+package com.graphhopper.routing.util.spatialrules;
+
+import org.locationtech.jts.geom.*;
+import org.locationtech.jts.geom.prep.PreparedGeometry;
+import org.locationtech.jts.geom.prep.PreparedGeometryFactory;
+import org.locationtech.jts.index.strtree.STRtree;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class AreaIndex<T extends AreaIndex.HasBorders> {
+
+    public interface HasBorders {
```

Comment:
An `AreaIndex` should IMHO keep an index of `Areas`:

```suggestion
    public interface Area {
```

Reply:
Thanks for reviewing this. I'm in the process of replacing spatial rules with custom areas, trying to find a good way to do so. Therefore there might be some back and forth changes.

Here I wasn't sure yet if we actually need to make the AreaIndex generic this way, but so far I think it has some advantages. We already use it for two different things: CustomAreas and the Landmark splitting areas.

Regarding the rename: I also like `Area`, so I left a little todo and we can decide later.